### ISSUE TITLE
Usage of yield

### ISSUE THREAD

**Issue by SOHELAHMED7** [yiisoft/yii2#16953]
This is for proposal/discussion/enh'ment

so feel free to share your views

I went through 
https://www.yiiframework.com/wiki/2541/when-to-use-active-record

then I thought this `yield`
http://php.net/manual/en/language.generators.syntax.php#control-structures.yield

can be used in CLI mode (cron, queue etc)

For eg
https://www.yiiframework.com/doc/api/2.0/yii-db-query#all()-detail

    public array all ( $db = null )

can be modified to 

    public array all ( $db = null, boolean $use_yield = false )

**Comment by samdark** [comment]
What's the benefit?

**Comment by SOHELAHMED7** [comment]
say for eg if a db table have 30+ cols and 100K records and you want to export in csv (single file) with the help of Yii active record

then

it will fetch one row by one but will not store in one single array of objects, and we can also use relations 


Imp. further ref

1. https://medium.com/tech-tajawal/use-memory-gently-with-yield-in-php-7e62e2480b8d

2. https://www.leaseweb.com/labs/2014/05/how-to-use-yield-keyword-php/

1st link contains benchmarking details and may explain better than me

**Comment by samdark** [comment]
Yes, I know what yield is for but I'm not sure it will give you enough benefits when used with AR.

**Comment by SOHELAHMED7** [comment]
Strong use case (for eg) is for  generating csv file.

    // case 1: all results (user) are stored in one variable, might consume more memory
    foreach(User::find()->all() as $key => $a_user) {
        // write $a_user data to csv file
    }

    // case 1: a result (user) is stored in one variable, consume less memory
    foreach(User::find()->all(null, true) as $key => $a_user) {
        // write $a_user data to csv file
    }

prevents php memory exhaustion issue 

IMHO, it will give benifits, I think

**Comment by SOHELAHMED7** [comment]
If yield is used then afaik  It can be only used as iterator for eg in foreach

Case 1 can be used as  $user[0]. That is the benefit of not using yield


Case 2 cannot be used as $user[0]. That is the limitation of using yield

**Comment by cebe** [comment]
> // case 1: a result (user) is stored in one variable, consume less memory

We already have `->each()` method on active query for that. it does not use yield, but it implements a similar pattern.

**Comment by SOHELAHMED7** [comment]
Agreed.
each() is similar to yield

**Comment by shiposhapv** [comment]
https://www.php.net/manual/en/mysqlinfo.concepts.buffering.php  it will help you)
